torch/nn/modules/module.py

```diff
@@ -500,32 +501,30 @@ def register_forward_hook(self, hook):
         self._forward_hooks[handle.id] = hook
         return handle
 
-    def _tracing_name(self, tracing_state):
-        if not tracing_state._traced_module_stack:
-            return None
-        module = tracing_state._traced_module_stack[-1]
-        for name, child in module.named_children():
-            if child is self:
-                return name
-        return None
 
     def _slow_forward(self, *input, **kwargs):
         tracing_state = torch._C._get_tracing_state()
-        if not tracing_state:
+        if not tracing_state or isinstance(self.forward, torch._C.ScriptMethod):
             return self.forward(*input, **kwargs)
-        if not hasattr(tracing_state, '_traced_module_stack'):
-            tracing_state._traced_module_stack = []
-        name = self._tracing_name(tracing_state)
-        if name:
-            tracing_state.push_scope('%s[%s]' % (self._get_name(), name))
-        else:
-            tracing_state.push_scope(self._get_name())
-        tracing_state._traced_module_stack.append(self)
+        recording_scopes = torch.jit._trace_module_map is not None
+        if recording_scopes:
+            name = torch.jit._trace_module_map[self] if self in torch.jit._trace_module_map else None
+            if name:
+                cur_scope_name = tracing_state.curr_scope()
+                if not name.startswith(cur_scope_name):
+                    warnings.warn('Called module {} from {}, but these modules are not in a '
```

Comment:
This restriction isn't necessary. The nesting of ScopeBlocks does not have to match the nesting of the module hierarchy. In the cases where it doesn't the module whose method _is_ being called is going to need to be passed in as a non-self argument to calling method. But this is something that is already handled by the capture process. Essentially this can be done with two steps:

1. Before converting subblocks, just add the correct `self` argument to the block, which is determined by the _trace_module_map.
2. Also before converting subblocks, replace all uses of variables inside the immediate subblock of values that are actually attributes of self for that subblock with their getattr equivalents. This would make sure we do not pass self.weight as an argument to the method but look it up from the module. This will also make all those `self` arguments we added in step 1 into get attrs on the parent when possible.

In cases where we do not have a hierarchy, then modules will get passed around as arguments that are not self, but the code will otherwise be correct.